### PR TITLE
jet-hadoop module improvements

### PR DESCRIPTION
**Pull request by cangencer**
* Additional javadocs added
* Added support for serializing some primitive types

### CODE REVIEW

File: hazelcast-jet-hadoop/src/main/java/com/hazelcast/jet/connector/hadoop/WritableSerializerHooks.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+        public boolean isOverwritable() {
+            return true;
+        }
+    }
+
+    private static class WritableStreamSerializer implements StreamSerializer<Writable> {
+
+        private static final ILogger LOGGER = Logger.getLogger(WritableStreamSerializer.class);
+
+        private boolean warnedOnce;
```
Comment (gurbuzali) [inline hazelcast-jet-hadoop/src/main/java/com/hazelcast/jet/connector/hadoop/WritableSerializerHooks.java]: how about concurrent read of this field ?

File: hazelcast-jet-hadoop/src/main/java/com/hazelcast/jet/connector/hadoop/WritableSerializerHook.java [hunk 2]
```diff
@@ -25,56 +24,58 @@
...
-public class WritableSerializerHook {
+/**
+ * This class can be extended to quickly create strongly typed serializers for
+ * {@code Writable} types
+ */
+public abstract class WritableSerializerHook<T extends Writable> implements SerializerHook<T> {
 
-    private static final int WRITABLE = -400;
+    private final Class<T> clazz;
+    private final Supplier<T> supplier;
```
Comment (mtopolnik) [inline hazelcast-jet-hadoop/src/main/java/com/hazelcast/jet/connector/hadoop/WritableSerializerHook.java]: `supplier` is a very generic name, I had to study all the code to figure out the purpose. Maybe `factory` is better, or `constructor` or something.

File: hazelcast-jet-hadoop/src/main/java/com/hazelcast/jet/connector/hadoop/WritableSerializerHook.java [hunk 3]
```diff
@@ -25,56 +24,58 @@
...
 
 import java.io.IOException;
+import java.util.function.Supplier;
 
-import static com.hazelcast.jet.impl.util.ExceptionUtil.rethrow;
-
-public class WritableSerializerHook {
+/**
+ * This class can be extended to quickly create strongly typed serializers for
+ * {@code Writable} types
```
Comment (mtopolnik) [inline hazelcast-jet-hadoop/src/main/java/com/hazelcast/jet/connector/hadoop/WritableSerializerHook.java]: I couldn't understand this doc. The class seems to support quick operation without the need to extend it and the actual subclasses provided don't extend anything. Is the requirement of the serializer hook mechanism to have distinct hook types?

File: hazelcast-jet-hadoop/src/main/java/com/hazelcast/jet/connector/hadoop/WriteHdfsP.java [hunk 4]
```diff
@@ -45,18 +46,32 @@
 import static org.apache.hadoop.mapreduce.TaskType.JOB_SETUP;
 
 /**
- * HDFS writer for Jet, consumes Map.Entry objects and writes them to the output file in HDFS.
+ * A processor which consumes records and writes them to HDFS.
+ *
+ * The records are written according to the given {@code OutputFormat}. The
+ * processor expects incoming items to be of the {@code Map.Entry<K,V>} and also
```
Comment (mtopolnik) [inline hazelcast-jet-hadoop/src/main/java/com/hazelcast/jet/connector/hadoop/WriteHdfsP.java]: "to be of the `Entry<K, V>` -> "to be of type `Entry<K, V>`"

File: hazelcast-jet-hadoop/src/main/java/com/hazelcast/jet/connector/hadoop/WriteHdfsP.java [hunk 5]
```diff
@@ -45,18 +46,32 @@
...
+ * accepts optional mappers to transform the key and the value to another type, such as
+ * a {@code Writable} that is expected by some of the output formats.
+ *
+ * Each processor instance creates a single file in the output path identified by
+ * the member ID and the processor ID. Unlike in MapReduce,
+ * the output files are not sorted by key.
+ *
+ * @param <K>         the key type of the records
```
Comment (mtopolnik) [inline hazelcast-jet-hadoop/src/main/java/com/hazelcast/jet/connector/hadoop/WriteHdfsP.java]: Sounds like `KR` :)

File: hazelcast-jet-hadoop/src/main/java/com/hazelcast/jet/connector/hadoop/WriteHdfsP.java [hunk 6]
```diff
@@ -45,18 +46,32 @@
...
+ * @param <KR>        the type of the key after mapping
```
Comment (mtopolnik) [inline hazelcast-jet-hadoop/src/main/java/com/hazelcast/jet/connector/hadoop/WriteHdfsP.java]: Maybe.... `KM`. Or maybe have `KI` and `KO`.

File: hazelcast-jet-hadoop/src/main/java/com/hazelcast/jet/connector/hadoop/WriteHdfsP.java [hunk 7]
```diff
@@ -88,51 +103,59 @@ public boolean isCooperative() {
     }
 
     /**
-     * Returns a meta-supplier of processors that write HDFS files.
+     * Returns a meta-supplier of processors that writes to HDFS. The processors expect
+     * items of type {@code Map.Entry<K,V>}. The key and the value must be of a type that
+     * is writable to disk by the given {@code OutputFormat} in the configuration.
+     *
+     * The supplied {@code JobConf} supplied must specify an {@code OutputFormat} with a path.
```
Comment (mtopolnik) [inline hazelcast-jet-hadoop/src/main/java/com/hazelcast/jet/connector/hadoop/WriteHdfsP.java]: Double "supplied"

File: hazelcast-jet-hadoop/src/main/java/com/hazelcast/jet/connector/hadoop/WritableSerializerHooks.java (hunk 1 above)
Comment (cangencer) [inline hazelcast-jet-hadoop/src/main/java/com/hazelcast/jet/connector/hadoop/WritableSerializerHooks.java]: Yes, I didn't want to do a volatile read for every time this method is called. It's not so important if there's more than one warning, as long as it doesn't flood the logs (which would happen if it warned everytime)

**Pull request merged**